ote/src/clj/ote/netex/netex_util.clj

```diff
@@ -0,0 +1,80 @@
+(ns ote.netex.netex_util
+  "Utilities for working with netex objects"
+  (:require [specql.core :as specql]
+            [specql.op :as op]
+            [ote.db.netex :as netex]
+            [ote.db.transport-service :as t-service]
+            ))
+
+(defn file-download-url [{{base-url :base-url} :environment} transport-service-id file-id]
+  (format "%sexport/netex/%d/%d" base-url transport-service-id file-id))
+
+(defn assoc-download-url [config conversions]
+  (map
+    #(if (= :ok (keyword (::netex/status %)))
+       (assoc %
+         :url
+         (file-download-url config
+                            (::netex/transport-service-id %)
+                            (::netex/id %)))
+       %)
+    conversions))
+
+(defn fetch-conversions-for-services
+  "Fetches successful netex conversion records for services in `service-id`s and appends each with a download :url key.
+  Returns a sequence of collections."
+  [config db service-ids]
+  (->> (specql/fetch db
+                     ::netex/netex-conversion
+                     (specql/columns ::netex/netex-conversion)
+                     (op/and
+                       {::netex/transport-service-id (op/in service-ids)}
+                       {::netex/status :ok}
+                       {::netex/filename op/not-null?}))
+       (assoc-download-url config)))
+
+(defn- append-ote-netex-url-to-interface
+  "Takes netex `conversions` and walks through `interfaces`, appending a netex download url to an interface which
+  matches to a conversion.
+  Return: `interfaces` where a netex download url is appended "
+  [config interfaces find-exif-id conversion]
+  (mapv (fn [exif]
+          (if (= (::t-service/id exif) find-exif-id)
+            (assoc exif
+              :url-ote-netex
+              (file-download-url config
+                                 (::netex/transport-service-id conversion)
+                                 (::netex/id conversion)))
+            exif))
+        interfaces))
+
+(defn append-ote-netex-urls
+  "`services` = collection of services
+  `config` = object which contains ote configuration properties for the environment
+  `db`= ote database
+  `ext-ifs-key` = A key, within a service object, which contains external interface objects collection
+  Return: collection of services with an ote-generated netex appended to those external interfaces for which a netex url is available."
+  [services config db ext-ifs-key]
+  (let [conversions (fetch-conversions-for-services config
+                                                    db
+                                                    (set (map #(::t-service/id %)
+                                                              services)))]
+    (reduce
```

Comment:
Just wondering if this could have been made with a simple for loop?

Reply:
Simple is always the best :)
What kind of loop did you have in mind?
Approach here was that `services` is not iterated because there might be a lot of them, and in any case typically there's less `conversions` than `services`. It seemed more optimal to go through the `conversions` and find the service matching to it.

From readability point of view iterating services in a `mapv` and in one such iteration iterating all `conversions` might be better. There's just the matter of possibility of several conversions per service, which has to be handled there as well.
I can try re-implementing this by having the main iteration to go through the `services`, if that was what you had in mind.

Reply:
I'm used to think that reduce is needed when the original map must be changed. e.g. split services to multiple different objects.
But in this case we need only to add item/items to external-interfaces vector that is already in the original map (services / single service)

Reply:
`reduce` passes state information from one iteration to another, the nature of state is up to the implementor. Transforming maps to other kind of maps is best done using map(v) function or `for`.
But that's besides the point, 
the more important design decision is which collection to iterate. 
This probably won't look any simpler if reduce is replaced using a mapv, the main problem will boil down to the nested iteration. I can do an alternative, but it might be you won't like it any better.

Reply:
> But in this case we need only to add item/items to external-interfaces vector that is already in the original map (services / single service)
That's not the case, here iteration should add a _key_ under some of the objects in the external-interfaces vector. No new items i.e. objects are added to the vector.

Reply:
If you don't mind, I'll merge this and let's check the improved loop together to avoid using time to something you don't want it to be used :)

Reply:
👍